src/drivers/ms5525_airspeed/MS5525.cpp

```diff
@@ -0,0 +1,331 @@
+			perf_count(_comms_errors);
+			return false;
+		}
+
+		// read 2 byte value
+		uint8_t val[2];
+		ret = transfer(nullptr, 0, &val[0], 2);
+
+		if (ret == PX4_OK) {
+			prom[i] = (val[0] << 8) | val[1];
+
+		} else {
+			perf_count(_comms_errors);
+			return false;
+		}
+	}
+
+	// Step 3 - check CRC
+	const uint8_t crc = prom_crc4(prom);
+	const uint8_t onboard_crc = prom[7] & 0xF;
+
+	if (crc == onboard_crc) {
+		// store valid calibration coefficients
+		C1 = prom[1];
+		C2 = prom[2];
+		C3 = prom[3];
+		C4 = prom[4];
+		C5 = prom[5];
+		C6 = prom[6];
+
+		Tref = C5 * (1UL << Q5);
+
+		return true;
+
+	} else {
+		PX4_ERR("CRC mismatch");
+		return false;
+	}
+}
+
+uint8_t
+MS5525::prom_crc4(uint16_t n_prom[]) const
+{
+	// see Measurement Specialties AN520
+
+	// crc remainder
+	unsigned int n_rem = 0x00;
+
+	// original value of the crc
+	unsigned int crc_read = n_prom[7]; // save read CRC
+	n_prom[7] = (0xFF00 & (n_prom[7])); // CRC byte is replaced by 0
+
+	// operation is performed on bytes
+	for (int cnt = 0; cnt < 16; cnt++) {
+		// choose LSB or MSB
+		if (cnt % 2 == 1) {
+			n_rem ^= (unsigned short)((n_prom[cnt >> 1]) & 0x00FF);
+
+		} else {
+			n_rem ^= (unsigned short)(n_prom[cnt >> 1] >> 8);
+		}
+
+		for (uint8_t n_bit = 8; n_bit > 0; n_bit--) {
+			if (n_rem & (0x8000)) {
+				n_rem = (n_rem << 1) ^ 0x3000;
+
+			} else {
+				n_rem = (n_rem << 1);
+			}
+		}
+	}
+
+	n_rem = (0x000F & (n_rem >> 12)); // final 4-bit reminder is CRC code
+	n_prom[7] = crc_read; // restore the crc_read to its original place
+
+	return (n_rem ^ 0x00);
+}
+
+int
+MS5525::collect()
+{
+	perf_begin(_sample_perf);
+
+	// read ADC
+	uint8_t cmd = CMD_ADC_READ;
+	int ret = transfer(&cmd, 1, nullptr, 0);
+
+	if (ret != PX4_OK) {
+		perf_count(_comms_errors);
+		return ret;
+	}
+
+	// read 24 bits from the sensor
+	uint8_t val[3];
+	ret = transfer(nullptr, 0, &val[0], 3);
+
+	if (ret != PX4_OK) {
+		perf_count(_comms_errors);
+		return ret;
+	}
+
+	uint32_t adc = (val[0] << 16) | (val[1] << 8) | val[2];
+
+	// If the conversion is not executed before the ADC read command, or the ADC read command is repeated, it will give 0 as the output
+	// result. If the ADC read command is sent during conversion the result will be 0, the conversion will not stop and
+	// the final result will be wrong. Conversion sequence sent during the already started conversion process will yield
+	// incorrect result as well.
+	if (adc == 0) {
+		perf_count(_comms_errors);
+		return EAGAIN;
+	}
+
+	if (_current_cmd == CMD_CONVERT_PRES) {
+		D1 = adc;
+		_pressure_count++;
+
+		if (_pressure_count > 9) {
+			_pressure_count = 0;
+			_current_cmd = CMD_CONVERT_TEMP;
+		}
+
+	} else if (_current_cmd == CMD_CONVERT_TEMP) {
+		D2 = adc;
+		_current_cmd = CMD_CONVERT_PRES;
+
+		// only calculate and publish after new pressure readings
+		return PX4_OK;
+	}
+
+	// not ready yet
+	if (D1 == 0 || D2 == 0) {
+		return EAGAIN;
+	}
+
+	// Difference between actual and reference temperature
```

Comment:
@tubeme from here down implements the calculation as described in your comment. https://github.com/PX4/Firmware/pull/7283#issuecomment-307374759